modules/idlgen/core/src/main/scala/converters.scala

```diff
@@ -0,0 +1,162 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package freestyle.rpc.idlgen.protobuf
+
+import freestyle.rpc.idlgen.protobuf.model._
+import freestyle.rpc.internal.ServiceAlg
+import scala.meta.Defn._
+import scala.meta._
+
+object converters {
+
+  trait ScalaMetaSource2ProtoDefinitions {
+    def convert(s: Source): ProtoDefinitions
+  }
+
+  class DefaultScalaMetaSource2ProtoDefinitions(
+      implicit OC: ScalaMetaObject2ProtoOptions,
+      MC: ScalaMetaClass2ProtoMessage,
+      SC: ScalaMetaTrait2ProtoService)
+      extends ScalaMetaSource2ProtoDefinitions {
+
+    override def convert(s: Source): ProtoDefinitions = ProtoDefinitions(
+      options = optionDirectives(s).toList.flatMap(OC.convert),
+      messages = messageClasses(s).map(MC.convert).toList,
+      services = serviceClasses(s).map(SC.convert).toList
+    )
+
+    private[this] def optionDirectives(source: Source): Seq[Object] = source.collect {
+      case o: Object /*if o.hasMod(mod"@option")*/ => o
+    }
+
+    private[this] def messageClasses(source: Source): Seq[Class] = source.collect {
+      case c: Class if c.mods.exists { case mod"@message" => true; case _ => false } => c
+    }
+
+    private[this] def serviceClasses(source: Source): Seq[Trait] = source.collect {
+      case t: Trait if t.mods.exists { case mod"@service" => true; case _ => false } => t
+    }
+  }
+
+  object ScalaMetaSource2ProtoDefinitions {
+    implicit def defaultSourceToProtoDefinitions(
+        implicit OC: ScalaMetaObject2ProtoOptions,
+        MC: ScalaMetaClass2ProtoMessage,
+        SC: ScalaMetaTrait2ProtoService): ScalaMetaSource2ProtoDefinitions =
+      new DefaultScalaMetaSource2ProtoDefinitions
+  }
+
+  trait ScalaMetaClass2ProtoMessage {
+    def convert(c: Class): ProtoMessage
+  }
+
+  object ScalaMetaClass2ProtoMessage {
+    implicit def defaultClass2MessageConverter(
+        implicit PC: ScalaMetaParam2ProtoMessageField): ScalaMetaClass2ProtoMessage =
+      new ScalaMetaClass2ProtoMessage {
+        override def convert(c: Class): ProtoMessage = ProtoMessage(
+          name = c.name.value,
+          fields = c.ctor.paramss.flatten.zipWithIndex.map {
+            case (p, t) => PC.convert(p, t + 1, None)
+          }.toList
+        )
+      }
+  }
+
+  trait ScalaMetaParam2ProtoMessageField {
+    def convert(p: Term.Param, tag: Int, mod: Option[ProtoFieldMod]): ProtoMessageField
+  }
+
+  object ScalaMetaParam2ProtoMessageField {
+    implicit def defaultParam2ProtoMessageField: ScalaMetaParam2ProtoMessageField =
+      new ScalaMetaParam2ProtoMessageField {
+        override def convert(
+            p: Term.Param,
+            tag: Int,
+            mod: Option[ProtoFieldMod] = None): ProtoMessageField = p match {
+          case param"..$mods $paramname: Double = $expropt" =>
```

Comment:
Given that in all of the pattern-matches below we are not using neither the `$paramname` nor the `mods` for choosing cases, we could just match on the `p.decltpe` instead.  
```Scala
p.decltpe match {
    case t"Float" => --- 
    case t"Long" => --- 
}
```
You can get the name of the parameter using `p.name`.

Reply:
Agreed, I'm working on a refactoring that addresses this and other issues. I think I'll make another PR for it.